### DISCUSSION TITLE
Support for building on Alpine?

### DISCUSSION THREAD

**Discussion by xplshn** [d99kris/nchat#198] [Ideas]
That's it, I want to propose either adding support for Alpine or publishing static releases, so that we can all try nchat.

**Comment by d99kris** [comment]
Hi @xplshn - yes it makes sense to at least build without issues on Alpine. I've prepared some local fixes for this, but I need to sanity test on a few more (non-alpine) platforms before I can publish a fix. Hope to get it done in the weekend. Will update here again.

As for providing builds, it's currently out of scope of the project. There are too many distros and architectures, but I have too little time. 😅 
If anyone would like to package nchat for any distro, it's very much welcome. 🙂

**Reply by xplshn** [reply]
Thanks!

**Comment by d99kris** [comment]
Hi - I've pushed [4b3c602](https://github.com/d99kris/nchat/commit/4b3c6022b81ef1cd513db93a7a09278d648a70f7) fixing nchat build on alpine linux. It should be possible to follow the [build from source](https://github.com/d99kris/nchat?tab=readme-ov-file#build-from-source) instructions, basically:
```
git clone https://github.com/d99kris/nchat && cd nchat
./make.sh deps
./make.sh build && ./make.sh install
```

**Reply by xplshn** [reply]
Okay, just another thing. Alpine comes with doas by default, on most Alpine setups people don't bother installing sudo because doas works just fine

**Reply by d99kris** [reply]
Thanks for sharing! I'll consider updating the make.sh script.

**Comment by xplshn** [comment]
And after changing sudo to doas in make.sh this happens:
```
]/Documents/ItALL/nchat @ ./make.sh build
-- Using -j1 (4 cores, 4000 MB phys mem, 3500 MB mem per core needed)
-- Using build type 'Release' (default).
-- Dummy: ON
-- Telegram: ON
-- WhatsApp: ON
-- Dynamic Load: ON
-- Multi Protocol: ON
-- Core Dump: ON
-- The C compiler identification is GNU 13.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - broken
CMake Error at /usr/share/cmake/Modules/CMakeTestCCompiler.cmake:67 (message):
  The C compiler

    "/usr/bin/cc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/home/anto/Documents/ItALL/nchat/build/CMakeFiles/CMakeScratch/TryCompile-pGNOLb'

    Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_2fdd7/fast
    /usr/bin/gmake  -f CMakeFiles/cmTC_2fdd7.dir/build.make CMakeFiles/cmTC_2fdd7.dir/build
    gmake[1]: Entering directory '/home/anto/Documents/ItALL/nchat/build/CMakeFiles/CMakeScratch/TryCompile-pGNOLb'
    Building C object CMakeFiles/cmTC_2fdd7.dir/testCCompiler.c.o
    /usr/bin/cc   -march=native -O2 -static-pie -static  -o CMakeFiles/cmTC_2fdd7.dir/testCCompiler.c.o -c /home/anto/Documents/ItALL/nchat/build/CMakeFiles/CMakeScratch/TryCompile-pGNOLb/testCCompiler.c
    Linking C executable cmTC_2fdd7
    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_2fdd7.dir/link.txt --verbose=1
    /usr/bin/cc -march=native -O2 -static-pie -static  -static  CMakeFiles/cmTC_2fdd7.dir/testCCompiler.c.o -o cmTC_2fdd7
    /usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: /usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/crtbeginT.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a PIE object
    /usr/lib/gcc/x86_64-alpine-linux-musl/13.2.1/../../../../x86_64-alpine-linux-musl/bin/ld: failed to set dynamic section sizes: bad value
    collect2: error: ld returned 1 exit status
    gmake[1]: *** [CMakeFiles/cmTC_2fdd7.dir/build.make:99: cmTC_2fdd7] Error 1
    gmake[1]: Leaving directory '/home/anto/Documents/ItALL/nchat/build/CMakeFiles/CMakeScratch/TryCompile-pGNOLb'
    gmake: *** [Makefile:127: cmTC_2fdd7/fast] Error 2





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  lib/ncutil/CMakeLists.txt:3 (project)


-- Configuring incomplete, errors occurred!
build failed, exiting.
]/Documents/ItALL/nchat @
```

**Reply by xplshn** [reply]
note for those that don't use/know alpine: Alpine Linux uses Musl as its libc

**Reply by d99kris** [reply]
I built nchat in Alpine Linux 3.19.1 on aarch64 as well as x86_64 - standard images, fresh installations. Perhaps our environments differ in some ways. For example I don't see `-march=native -O2 -static-pie -static` when `testCCompiler.c` is built testing the C compiler on my Alpine systems. Could you please help check if you have some special compiler options set up?
```
echo $CC
echo $CXX
echo $CFLAGS
echo $CXXFLAGS
echo $CPPFLAGS
echo $LDFLAGS
```

**Reply by xplshn** [reply]
Doubt -march is the problem, but I forgot I had -static set as ldflags. WIll check again
```

]/Documents/ItALL @ echo $CC

]/Documents/ItALL @ echo $CXX

]/Documents/ItALL @ echo $CFLAGS
-march=native -O2 -static-pie -static
]/Documents/ItALL @ echo $CXXFLAGS

]/Documents/ItALL @ echo $CPPFLAGS

]/Documents/ItALL @ echo $LDFLAGS
-static
]/Documents/ItALL @

```

**Reply by xplshn** [reply]
Oh, now I have a different issue. I unset every var you mentioned.
```
/home/anto/Documents/ItALL/nchat/src/uicolorconfig.cpp: In static member function 'static int UiColorConfig::GetAttribute(const std::string&)':
/home/anto/Documents/ItALL/nchat/src/uicolorconfig.cpp:199:17: error: 'A_ITALIC' was not declared in this scope
  199 |     { "italic", A_ITALIC },
      |                 ^~~~~~~~
/home/anto/Documents/ItALL/nchat/src/uicolorconfig.cpp:200:3: error: could not convert '{{"normal", 0}, {"underline", 512}, {"reverse", 1024}, {"bold", 8192}, {"italic", <expression error>}}' from '<brace-enclosed initializer list>' to 'const std::map<std::__cxx11::basic_string<char>, int>'
  200 |   };
      |   ^
      |   |
      |   <brace-enclosed initializer list>
make[2]: *** [CMakeFiles/nchat.dir/build.make:104: CMakeFiles/nchat.dir/src/uicolorconfig.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:475: CMakeFiles/nchat.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
build failed, exiting.
]/Documents/ItALL/nchat @
```

**Reply by d99kris** [reply]
Hi, thanks for reporting. I've pushed 731aa92 to fix the A_ITALIC issue, it appears it's a non-standard ncurses extension that applications need to check for.

**Reply by xplshn** [reply]
Sorry to keep bothering. Maybe you should try building in Alpine x86 rather than arm? Issues keep popping up, I already unset the vars, deleted ./build and re ran ./make.sh
```
[  7%] Building CXX object CMakeFiles/nchat.dir/src/uikeyconfig.cpp.o
/home/anto/Documents/ItALL/nchat/src/uikeyconfig.cpp: In static member function 'static int UiKeyConfig::GetVirtualKeyCodeFromOct(const std::string&)':
/home/anto/Documents/ItALL/nchat/src/uikeyconfig.cpp:337:26: error: invalid conversion from 'const char*' to 'char*' [-fpermissive]
  337 |   define_key(keyStr.c_str(), keyCode);
      |              ~~~~~~~~~~~~^~
      |                          |
      |                          const char*
In file included from /home/anto/Documents/ItALL/nchat/src/uikeyconfig.cpp:12:
/usr/local/include/curses.h:669:21: note:   initializing argument 1 of 'int define_key(char*, int)'
  669 | int      define_key(char *, int);
      |                     ^~~~~~
make[2]: *** [CMakeFiles/nchat.dir/build.make:244: CMakeFiles/nchat.dir/src/uikeyconfig.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:475: CMakeFiles/nchat.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
build failed, exiting.
]/Documents/ItALL/nchat @
```

**Reply by d99kris** [reply]
Actually...

> I built nchat in Alpine Linux 3.19.1 on aarch64 as well as x86_64

I suspect maybe we have different curses packages (or versions). Mine was installed using:

```apk add ncurses-dev```

If I list `curses` related packages I get this:
```
$ apk info -vv | grep curses
libncurses++-6.4_p20231125-r0 - Console display library (libncurses++)
libncursesw-6.4_p20231125-r0 - Console display library (libncursesw)
ncurses-dev-6.4_p20231125-r0 - Console display library (development files)
ncurses-terminfo-base-6.4_p20231125-r0 - Descriptions of common terminals
```

Perhaps you could help check what curses related packages you have or which installed package is providing your `/usr/local/include/curses.h` (this file is not present on neither my aarch64 nor x86_64 system), maybe something like:
```
apk info --who-owns /usr/local/include/curses.h
```

**Comment by d99kris** [comment]
It should be possible to build on Alpine, however WhatsApp protocol does not work due to an incompatibility between Go and musl - see https://github.com/d99kris/nchat/issues/204 for details. I will proceed to close this issue. Feel free to re-open or report a new bug if there are build issues on Alpine. Thanks!

**Comment by d99kris** [comment]
Pre-built binaries has been added for the latest release [v5.17.26](https://github.com/d99kris/nchat/releases/tag/v5.17.26) with support for Alpine (musl builds). For these builds a patched version of Go is used, enabling WhatsApp and Signal libraries to work with musl.

For installation steps see [Install Pre-built Binary](https://github.com/d99kris/nchat#install-pre-built-binary).
